Add function `seriesDecomposeSTL()` for time series decomposition into seasonal, trend and residual components

bhavnajindal opened the pull request:
Implemented function `seriesDecomposeSTL()` for seasonal decompose of time series data. This function uses the (MIT-licensed) [stl-cpp](https://github.com/ankane/stl-cpp) library.

Decomposing a time series into seasonal, trend, and residual components is a statistical method used to analyze data that changes over time. This method can be used to make informed decisions about their services by identifying patterns in data and understanding its overall behaviour. Applications of `seriesDecomposeSTL()` include:
- Forecasting
- Anomly detection
- Outliers

Parameters: An array containing time series data and the period of the time series.

Returns: An array of arrays where the first array include seasonal components, the second array contains a trend,
and the third array a residual component. 

Example:

```
SELECT seriesDecomposeSTL([10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34], 3);
```

Output:
```
[[-13.529999,-3.1799996,16.71,-13.53,-3.1799996,16.71,-13.53,-3.1799996,16.71,-13.530001,-3.18,16.710001,-13.530001,-3.1800003,16.710001,-13.530001,-3.1800003,16.710001,-13.530001,-3.1799994,16.71,-13.529999,-3.1799994,16.709997],[23.63,23.63,23.630003,23.630001,23.630001,23.630001,23.630001,23.630001,23.630001,23.630001,23.630001,23.63,23.630001,23.630001,23.63,23.630001,23.630001,23.63,23.630001,23.630001,23.630001,23.630001,23.630001,23.630003],[0,0.0000019073486,-0.0000019073486,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-0.0000019073486,0,0]]
```

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added function `seriesDecomposeSTL()` which decomposes a time series into a season, a trend and a residual component.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

bhavnajindal committed:
Added new library stl-cpp for sesonal decompose of time series data

bhavnajindal committed:
implemneted seriesDecomposeSTL method

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #57078 from ClibMouse/ClickHouse:seriesDecompose -->
*This is an automated comment for commit fd5e00468399463692652177c734f3a4162c99cd with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57078/fd5e00468399463692652177c734f3a4162c99cd/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6949284046">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57078/fd5e00468399463692652177c734f3a4162c99cd/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

bhavnajindal committed:
fixed the failing fasttest and style-check

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -0,0 +1,61 @@
+---
```
This file is shown as newly introduced but I think you want to merge from master and add new docs to the existing markdown added with https://github.com/ClickHouse/ClickHouse/pull/56171

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -0,0 +1,61 @@
+
+## seriesDecomposeSTL
+
+Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
```
Please link STL, e.g. to https://www.wessa.net/download/stl.pdf if that's what you had in mind.

rschu1ze replied:
Please use proper markdown link syntax:

```md
[...] based on [STL (Seasonal-Trend Decomposition Procedure Based on Loess)](https://www.wessa.net/download/stl.pdf) [...]
```

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -0,0 +1,61 @@
+## seriesDecomposeSTL
+
+Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
+Returns an array of arrays where the first array include seasonal components, the second array - trend,
```
Returns an array of three arrays [...]

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,183 @@
+#include "config.h"
+
+#if USE_SEASONAL
+#    ifdef __clang__
```
(cosmetic: let's remove the indentation on l. 4-25, preprocessor directives (l. 3) are no language constructs like `if`, `for` etc., and therefore don't affect indentation)

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,183 @@
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    bool isVariadic() const override { return true; }
```
Why variadic?

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,183 @@
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeArray>(std::make_shared<DataTypeFloat32>()));
```
So if the input is an array of Float64, will the output be Float32-based? Thats a loss of precision.

bhavnajindal replied:
The `str-cpp` library currents supports upto float32 precision only.

rschu1ze commented on src/configure_config.cmake:
```diff
@@ -160,6 +160,9 @@ endif ()
 if (ENABLE_OPENSSL)
     set(USE_OPENSSL_INTREE 1)
 endif ()
+if (TARGET ch_contrib::stl-cpp)
```
"std-cpp" is too easy to confuse with C++'s STL. Maybe call this "ch_contrib::seasonal_decompose" or something similarly verbose

bhavnajindal committed:
handled edge cases and addresses review comments

bhavnajindal commented:
Decomposing a time series into seasonal, trend, and residual components is a statistical method used to analyze data that changes over time. This method can be used by customers/businesses to make informed decisions about their services by identifying patterns in data and understanding its overall behaviour. The applications of series_decompose include:

- Forecasting
- Anomly detection
- Outliers

The newer dependency `stl-cpp` is a header only, lightweight library that uses `Seasonal and Trend decomposition using Loess` to compute the above components.

bhavnajindal committed:
handled NaN and inf values

jkartseva commented:
My question is whether this use case is broad enough to be considered for a CH function and what the anticipated number of potentially interested customers might be. If the application is too niche, I believe we should try another workflow? For instance, exporting data and applying an external library specializing in statistics, e.g. SciPy and [statsmodels](https://github.com/statsmodels/statsmodels/) (the latter also supports [STL decomposition](https://www.statsmodels.org/devel/examples/notebooks/generated/stl_decomposition.html))."
I'll bring this up for internal discussion. This is the 2nd PR introducing a dependency in a form of submodule. Will more follow, and if so, which are the anticipated dependencies or features?

rschu1ze commented:
The main concern is about
- introducing obscure/unmaintained/buggy 3rd party dependencies and/or
- obscure functions that will be relevant to only a very small subset of users.

Anyways, I took a look and season and trend decomposition using Loess (STL) seems to be quite popular ([Hyndman's popular textbook](https://otexts.com/fpp2/stl.html) references it which is a good sign, also the [original paper](http://www.nniiem.ru/file/news/2016/stl-statistical-model.pdf) has 3800 citations which proves that it is significant. The [C++ library](https://github.com/ankane/stl-cpp) was written by just one author (bad sign), but the code itself is not terribly complex (mostly procedural, good sign) and it is self-contained (header-only, good sign). That it is needed by KQL is also a good sign. So, green light from my side (though others may weigh in)

bhavnajindal committed:
remove stl-cpp submodule

bhavnajindal committed:
Added additional testcases

bhavnajindal committed:
fix style-check

bhavnajindal commented:
- Removed submodule dependency, instead copied str-cpp as suggested in comments. 
- Added test cases for multi-row inputs.

@jkartseva - Please review.

bhavnajindal committed:
fix for const column input

bhavnajindal committed:
fix testcase

bhavnajindal commented:
- Removed submodule dependency and copied stl-cpp as recommended in review 
- Handled multi-row input

@jkartseva - Please review

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -4,7 +4,7 @@ sidebar_position: 172
 ---
 
-# Time Series Functions
+#Time Series Functions
```
Cosmetic: Let's not remove spaces needlessly.

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -57,3 +57,58 @@ Result:
+
+## seriesDecomposeSTL
+
+Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
```
`Decomposes a time series using STL (Seasonal-Trend Decomposition Procedure Based on Loess)](https://www.wessa.net/download/stl.pdf) into a season, a trend and a residual component.'

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -57,3 +57,58 @@ Result:
+## seriesDecomposeSTL
+
+Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
+Returns an array of three arrays where the first array include seasonal components, the second array - trend,
```
The specifics of what is returned (l. 64-65) can be moved to the `**Returned value**` sub-section below.

rschu1ze commented on src/Common/config.h.in:
```diff
@@ -59,6 +59,7 @@
 #cmakedefine01 USE_PRQL
 #cmakedefine01 USE_OPENSSL_INTREE
 #cmakedefine01 USE_ULID
+#cmakedefine01 USE_SEASONAL
```
The preprocessor symbols are needed only for 3rd party libraries. Since you technically wrote your own STL implementation, l. 62 can be deleted.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+#include "config.h"
```
L. 1 can be removed, see my comment in config.h.in.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"time-series", &isArray<IDataType>, nullptr, "Array"},
```
Minor:`time-series` --> `time_series`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+        const ColumnArray * array = checkAndGetColumn<ColumnArray>(array_ptr.get());
+        if (!array)
+        {
+            const ColumnConst * const_array = checkAndGetColumnConst<ColumnArray>(
```
Cosmetic: Let's remove the newline between l. 71/72.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+
+        Float64 period;
+
+        auto ret = ColumnFloat32::create();
```
`ret` --> `res`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+        for (size_t i = 0; i < src_offsets.size(); ++i)
+        {
+            auto period_ptr = arguments[1].column->convertToFullColumnIfConst();
+            if (checkAndGetColumn<ColumnUInt8>(period_ptr.get()) || checkAndGetColumn<ColumnUInt16>(period_ptr.get())
```
A single `checkAndGetColumn<...>()` per line would be slightly easier to read.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+            else if (checkAndGetColumn<ColumnFloat32>(period_ptr.get()) || checkAndGetColumn<ColumnFloat64>(period_ptr.get()))
+            {
+                period = period_ptr->getFloat64(i);
+                if (isNaN(period) || !std::isfinite(period) || period < 0)
```
I am pretty sure we can use `std::isnan()` directly instead of `isNaN()` (period is already a FP-value and no fallback handling for non-FP types is needed).

rschu1ze replied:
`period < 0.0`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+        auto ret = ColumnFloat32::create();
+        auto & res_data = ret->getData();
+
+        ColumnArray::ColumnOffsets::MutablePtr col_offsets = ColumnArray::ColumnOffsets::create();
```
`col_offsets` and `col_offsets_data` are returned as result, right? Could they have `res` in their names?

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+        size_t len = end - start;
+        if (len < 4)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "At least four data points are needed for function {}", getName());
+        else if (period > (len / 2.0))
```
`else if` --> `if`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS, "The series should have data of at least two period lengths for function {}", getName());
+
+        std::vector<float> src((src_vec.begin() + start), (src_vec.begin() + end));
```
Cosmetic: Inner parentheses are not necessary: `std::vector<float> src(src_vec.begin() + start, src_vec.begin() + end);`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+
+        try
+        {
+            auto res = stl::params().fit(src, static_cast<size_t>(std::round(period)));
```
As `fit()` accepts only size_t-valued periods and we are doing rounding, we should restrict the period argument of SQL function `seriesDecomposeSTL` to only integers in the first place.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+            if (res.seasonal.empty())
+                return false;
+
+            seasonal = res.seasonal;
```
l. 195-197 are doing copies each. You might want to do this instead:

```cpp
seasonal = std::move(res.seasonal);
[...]
```

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,264 @@
+            residue = res.remainder;
+            return true;
+        }
+        catch (...)
```
Exception-to-error-code translation (or even swallowing exceptions) is a code smell. Let's remove the try/catch and have exceptions bubble up to the caller.

rschu1ze commented on src/Functions/stl.hpp:
```diff
@@ -0,0 +1,510 @@
+/*!
```
Future readers will wonder where this file comes from, what it does, if it is part of ClickHouse and if it isn't, what are the changes to the original version.

Therefore, let's add a comment:
```cpp
/// Dump of https://github.com/ankane/stl-cpp/blob/f3e1cd02b4f0341a7539ed1a171b21b14b37ed51/include/stl.hpp. Added to ClickHouse source code and not referenced as a submodule because of <Reason 1> and <Reason 2>.
```

(please check that the link points to the version you used and fill in the reasons, thanks!)

rschu1ze commented on tests/queries/0_stateless/02813_series_decompose.sql:
```diff
@@ -0,0 +1,21 @@
+-- Tags: no-fasttest, no-cpu-aarch64
```
`no-fasttest` is obsolete by now and can be removed as we are no longer using a submodule

rschu1ze replied:
Could we rename the test + reference file to `02813_seriesDecomposeSTL.sql`?

rschu1ze commented on tests/queries/0_stateless/02813_series_decompose.sql:
```diff
@@ -0,0 +1,21 @@
+
+DROP TABLE IF EXISTS tb2;
+
+CREATE TABLE tb2 (`n` UInt32, `a` Array(Float64)) ENGINE = Memory;
```
`n` --> id, `a` --> ts (aka. timeseries)

rschu1ze commented on tests/queries/0_stateless/02813_series_decompose.sql:
```diff
@@ -0,0 +1,21 @@
+INSERT INTO tb2 VALUES (14, [139, 87, 110, 68, 54, 50, 51, 53, 133, 86, 141, 97, 156, 94, 149, 95, 140, 77, 61, 50, 54, 47, 133, 72, 152, 94, 148, 105, 162, 101, 160, 87, 63, 53, 55, 54, 151, 103, 189, 108, 183, 113, 175, 113, 178, 90, 71, 62, 62, 65, 165, 109, 181, 115, 182, 121, 178, 114, 170]);
+
+SELECT seriesDecomposeSTL([10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34, 10.1, 20.45, 40.34], 3);
+SELECT seriesDecomposeSTL([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2], 0);
```
Period = 0 sounds funny. How is that supposed to work?

bhavnajindal replied:
When a `signal`/time-series is flat( as in case of [2,2,2,2,2,2,2,2,2,2,2,2,2,2]), its period is 0. This is just an edge testcase.

bhavnajindal committed:
changed 2nd argument type to UInt and other cosmetic changes

rschu1ze commented on docs/en/sql-reference/functions/time-series-functions.md:
```diff
@@ -57,3 +57,56 @@ Result:
+**Arguments**
+
+- `series` - An array of numeric values
+- `period` - A positive number
```
`number` --> `integer`

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+#pragma clang diagnostic pop
+#endif
+
+#include <cmath>
```
This include and the one on l. 23 are obsolete.

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+extern const int ILLEGAL_COLUMN;
+}
+
+/*
```
Cosmetic:

```cpp
/// Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
```

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+        const IColumn & src_data = array->getData();
+        const ColumnArray::Offsets & src_offsets = array->getOffsets();
+
+        UInt64 period;
```
Let's move that into the loop:

```cpp
        for (size_t i = 0; i < src_offsets.size(); ++i)
        {
            UInt64 period;

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+                || executeNumber<Float32>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue)
+                || executeNumber<Float64>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue))
+            {
+                res_data.insert(res_data.end(), seasonal.begin(), seasonal.end());
```
I am pretty sure that we could do

```cpp
res_data.insert(seasonal.begin(), seasonal.end());
```

instead (please test it).

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+    bool executeNumber(
+        const IColumn & src_data,
+        UInt64 period,
+        ColumnArray::Offset & start,
```
Let's pass `start` and `end` by value (instead of by reference)

rschu1ze commented on src/Functions/seriesDecomposeSTL.cpp:
```diff
@@ -0,0 +1,243 @@
+
+        std::vector<float> src(src_vec.begin() + start, src_vec.begin() + end);
+
+        auto res = stl::params().fit(src, static_cast<size_t>(period));
```
The cast is not necessary, `size_t` is on all major platforms defined as `UInt64`.

rschu1ze reviewed (approved):
I wanted to push some minor fixups directly but noticed that I did not have permissions to push to `ClibMouse` directly. Posting the comments here instead. After that, we can merge, thanks again!

bhavnajindal committed:
minor refactoring and cosmetic changes
